x/bep3/keeper/swap.go

```diff
@@ -0,0 +1,340 @@
+		return types.ErrAtomicSwapNotFound(k.codespace, swapID)
+	}
+
+	// Only unexpired AtomicSwaps can receive deposits
+	if uint64(ctx.BlockHeight()) > atomicSwap.ExpirationBlock {
+		return types.ErrAtomicSwapHasExpired(k.codespace)
+	}
+
+	atomicSwapCoin := atomicSwap.Amount[0]
+	coin := coins[0]
+
+	// Validate new deposit
+	if atomicSwap.CrossChain {
+		return types.ErrOnlySameChain(k.codespace)
+	}
+	if !atomicSwap.From.Equals(from) {
+		return types.ErrOnlyOriginalCreator(k.codespace, from, atomicSwap.From)
+	}
+	if atomicSwapCoin.Denom != coin.Denom {
+		return types.ErrInvalidCoinDenom(k.codespace, atomicSwapCoin.Denom, coin.Denom)
+	}
+
+	// Send coins from depositor to the bep3 module
+	err = k.supplyKeeper.SendCoinsFromAccountToModule(ctx, from, types.ModuleName, coins)
+	if err != nil {
+		return sdk.ErrInternal(err.Error())
+	}
+
+	ctx.EventManager().EmitEvent(
+		sdk.NewEvent(
+			types.EventTypeDepositAtomicSwap,
+			sdk.NewAttribute(types.AttributeKeyAtomicSwapID, fmt.Sprintf("%s", atomicSwap.SwapID)),
+			sdk.NewAttribute(types.AttributeKeyCoinDenom, fmt.Sprintf("%s", coin.Denom)),
+			sdk.NewAttribute(types.AttributeKeyCoinAmount, fmt.Sprintf("%d", coin.Amount.Int64())),
+		),
+	)
+
+	// Update AtomicSwap state
+	atomicSwap.Amount = atomicSwap.Amount.Add(coins)
+	currExpectedIncome, _ := sdk.ParseCoins(atomicSwap.ExpectedIncome)
+	atomicSwap.ExpectedIncome = currExpectedIncome.Add(coins).String()
+
+	k.SetAtomicSwap(ctx, atomicSwap)
+
+	return nil
+}
+
+// ClaimAtomicSwap validates a claim attempt, and if successful, sends the escrowed amount and closes the AtomicSwap
+func (k Keeper) ClaimAtomicSwap(ctx sdk.Context, from sdk.AccAddress, swapID []byte, randomNumber []byte) sdk.Error {
+
+	atomicSwap, found := k.GetAtomicSwap(ctx, swapID)
+	if !found {
+		return types.ErrAtomicSwapNotFound(k.codespace, swapID)
+	}
+
+	// Only unexpired AtomicSwaps can be claimed
+	if uint64(ctx.BlockHeight()) > atomicSwap.ExpirationBlock {
+		return types.ErrAtomicSwapHasExpired(k.codespace)
+	}
+
+	//  Calculate hashed secret using submitted number
+	hashedSubmittedNumber := types.CalculateRandomHash(randomNumber, atomicSwap.Timestamp)
+	hashedSecret, err := types.CalculateSwapID(hashedSubmittedNumber, atomicSwap.From, atomicSwap.SenderOtherChain)
+	if err != nil {
+		return sdk.ErrInternal(err.Error())
+	}
+
+	if !bytes.Equal(hashedSecret, swapID) {
+		return types.ErrInvalidClaimSecret(k.codespace, hashedSecret, swapID)
+	}
+
+	// If AtomicSwap is not cross-chain, atomicSwap.ExpectedIncome equals atomicSwap.Amount
+	claimerCoins, err := sdk.ParseCoins(atomicSwap.ExpectedIncome)
+	if err != nil {
+		return sdk.ErrInternal(err.Error())
+	}
+
+	if atomicSwap.CrossChain {
+		err := k.ValidateCoinMint(ctx, atomicSwap.Amount)
+		if err != nil {
+			return err
+		}
+
+		// Mint full amount of this coin's associated debt coin to bep3 module for internal limit tracking
+		internalTrackingCoins, err2 := getEqualInternalTrackingCoins(atomicSwap.Amount)
```

Comment:
I'm a little confused how the internal tracking coin is used. It seems like it could be replaced by a store value of `TotalSwapCoins` for each denom, and then just check that the requested swap doesn't put the total over the limit. Is there a reason to do it using coins and not just a counter?

Reply:
Not sure if it helps in this case, but supply has methods that return total amount of any coin.

Reply:
Added PrefixAssetSupplyKey to the store in https://github.com/Kava-Labs/kava/pull/370/commits/40edb8c627ef547d0f67fc0fb5559385d250cd05. This has the additional benefit of allowing us to import/export genesis coin supply amounts.